File: content/09-backtracking/permutations.js
/**
 * Problem: Print all Permutations of String
 * GeeksforGeeks: https://practice.geeksforgeeks.org/problems/permutations-of-a-given-string/0
 */

function findPermutations(s) {
    /**
     * Time Complexity: O(N * N!)
     * Space Complexity: O(N * N!)
     */
    const res = new Set();
    const chars = s.split('');
    
    function backtrack(l) {
        if (l === chars.length) {
            res.add(chars.join(''));
            return;
        }
        for (let i = l; i < chars.length; i++) {
            // Swap
            [chars[l], chars[i]] = [chars[i], chars[l]];
            backtrack(l + 1);
            // Backtrack swap
            [chars[l], chars[i]] = [chars[i], chars[l]];
        }
    }
    
    backtrack(0);
    return Array.from(res).sort();
}

if (require.main === module) {
    const assert = require('assert');
    assert.deepStrictEqual(findPermutations("ABC"), ["ABC", "ACB", "BAC", "BCA", "CAB", "CBA"]);
    assert.deepStrictEqual(findPermutations("AB"), ["AB", "BA"]);
    assert.deepStrictEqual(findPermutations("ABA"), ["AAB", "ABA", "BAA"]);
    console.log("JS: findPermutations passed tests!");
}

File: content/09-backtracking/permutations.py
"""
Problem: Print all Permutations of String
GeeksforGeeks: https://practice.geeksforgeeks.org/problems/permutations-of-a-given-string/0
"""

def find_permutations(s: str) -> list[str]:
    """
    Time Complexity: O(N * N!)
    Space Complexity: O(N * N!) for outputs and O(N) recursion stack
    """
    res = set()
    chars = list(s)
    
    def backtrack(l):
        if l == len(chars):
            res.add("".join(chars))
            return
            
        for i in range(l, len(chars)):
            chars[l], chars[i] = chars[i], chars[l]
            backtrack(l + 1)
            chars[l], chars[i] = chars[i], chars[l] # backtrack
            
    backtrack(0)
    return sorted(list(res))

if __name__ == "__main__":
    assert find_permutations("ABC") == ["ABC", "ACB", "BAC", "BCA", "CAB", "CBA"]
    assert find_permutations("AB") == ["AB", "BA"]
    assert find_permutations("ABA") == ["AAB", "ABA", "BAA"]
    print("Python: find_permutations passed tests!")
